// twilio/lib/rest/insights/v3/metadata.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio - Insights
 * Insights Domain V3 API.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataInstance = exports.InsightsMetadataResponseCubesMeasures = exports.InsightsMetadataResponseCubesDimensions = exports.InsightsMetadataResponseCubes = void 0;
exports.MetadataListInstance = MetadataListInstance;
const util_1 = require("util");
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
class InsightsMetadataResponseCubes {
    constructor(payload) {
        this.name = payload["name"];
        this.description = payload["description"];
        this.measures = payload["measures"];
        this.dimensions = payload["dimensions"];
    }
}
exports.InsightsMetadataResponseCubes = InsightsMetadataResponseCubes;
class InsightsMetadataResponseCubesDimensions {
    constructor(payload) {
        this.name = payload["name"];
        this.description = payload["description"];
        this.type = payload["type"];
    }
}
exports.InsightsMetadataResponseCubesDimensions = InsightsMetadataResponseCubesDimensions;
class InsightsMetadataResponseCubesMeasures {
    constructor(payload) {
        this.name = payload["name"];
        this.description = payload["description"];
        this.type = payload["type"];
        this.aggregation = payload["aggregation"];
    }
}
exports.InsightsMetadataResponseCubesMeasures = InsightsMetadataResponseCubesMeasures;
function MetadataListInstance(version) {
    const instance = {};
    instance._version = version;
    instance._solution = {};
    instance._uri = `/InsightsDomains/Conversations/Metadata`;
    instance.fetch = function fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new MetadataInstance(operationVersion, payload));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.fetchWithHttpInfo = function fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new MetadataInstance(operationVersion, response.body),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
/**
 * Response containing metadata about available cubes, measures, and dimensions for a domain
 */
class MetadataInstance {
    constructor(_version, _payload) {
        this._version = _version;
        const payload = _payload;
        this.domain = payload.domain;
        this.cubes =
            payload.cubes !== null && payload.cubes !== undefined
                ? payload.cubes.map((payload) => new InsightsMetadataResponseCubes(payload))
                : null;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            domain: this.domain,
            cubes: this.cubes,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.MetadataInstance = MetadataInstance;
