// twilio/lib/rest/memory/v1/dataMapping.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Memory API
 * APIs for managing memory stores, profiles, events, and conversational intelligence capabilities.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataMappingPage = exports.DataMappingInstance = exports.DataMappingContextImpl = exports.MappingTraitItem = exports.DataMappingToTraits = exports.DataMappingFromDataSet = exports.DataMappingCore = exports.CreateDataMappingInput = void 0;
exports.DataMappingListInstance = DataMappingListInstance;
const util_1 = require("util");
const TokenPage_1 = __importDefault(require("../../../base/TokenPage"));
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
/**
 * A data mapping connects external data sources to a Memory Store, enabling automatic ingestion and mapping of data to profile traits.
 */
class CreateDataMappingInput {
    constructor(payload) {
        this.displayName = payload["displayName"];
        this.description = payload["description"];
        this.isEnabled = payload["isEnabled"];
        this.mappingTo = payload["mappingTo"];
        this.mappingFrom = payload["mappingFrom"];
    }
}
exports.CreateDataMappingInput = CreateDataMappingInput;
/**
 * Writable fields of a data mapping. Used directly as the PATCH request body (all fields optional). Composed into CreateDataMappingInput via allOf.
 */
class DataMappingCore {
    constructor(payload) {
        this.displayName = payload["displayName"];
        this.description = payload["description"];
        this.isEnabled = payload["isEnabled"];
        this.mappingTo = payload["mappingTo"];
    }
}
exports.DataMappingCore = DataMappingCore;
/**
 * Configuration for mapping from a TDI dataset.
 */
class DataMappingFromDataSet {
    constructor(payload) {
        this.type = payload["type"];
        this.datasetId = payload["datasetId"];
    }
}
exports.DataMappingFromDataSet = DataMappingFromDataSet;
/**
 * Configuration for mapping data to traits in a Memory Store.
 */
class DataMappingToTraits {
    constructor(payload) {
        this.type = payload["type"];
        this.mappings = payload["mappings"];
    }
}
exports.DataMappingToTraits = DataMappingToTraits;
/**
 * Maps a field from a source to a Trait in the Memory Store.
 */
class MappingTraitItem {
    constructor(payload) {
        this.fieldName = payload["fieldName"];
        this.expression = payload["expression"];
        this.traitGroup = payload["traitGroup"];
        this.traitName = payload["traitName"];
    }
}
exports.MappingTraitItem = MappingTraitItem;
class DataMappingContextImpl {
    constructor(_version, storeId, dataMappingId) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(storeId)) {
            throw new Error("Parameter 'storeId' is not valid.");
        }
        if (!(0, utility_1.isValidPathParam)(dataMappingId)) {
            throw new Error("Parameter 'dataMappingId' is not valid.");
        }
        this._solution = { storeId, dataMappingId };
        this._uri = `/ControlPlane/Stores/${storeId}/DataMappings/${dataMappingId}`;
    }
    remove(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "delete",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new DataMappingInstance(operationVersion, payload, instance._solution.storeId, instance._solution.dataMappingId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    removeWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // DELETE operation that returns a response model
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "delete",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new DataMappingInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.dataMappingId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new DataMappingInstance(operationVersion, payload, instance._solution.storeId, instance._solution.dataMappingId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new DataMappingInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.dataMappingId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patch(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.patch({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new DataMappingInstance(operationVersion, payload, instance._solution.storeId, instance._solution.dataMappingId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patchWithHttpInfo(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .patchWithResponseInfo({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new DataMappingInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.dataMappingId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.DataMappingContextImpl = DataMappingContextImpl;
class DataMappingInstance {
    constructor(_version, _payload, storeId, dataMappingId) {
        this._version = _version;
        const payload = _payload;
        this.message = payload.message;
        this.statusUrl = payload.statusUrl;
        this.displayName = payload.displayName;
        this.description = payload.description;
        this.isEnabled = payload.isEnabled;
        this.mappingTo =
            payload.mappingTo !== null && payload.mappingTo !== undefined
                ? new DataMappingToTraits(payload.mappingTo)
                : null;
        this.mappingFrom =
            payload.mappingFrom !== null && payload.mappingFrom !== undefined
                ? new DataMappingFromDataSet(payload.mappingFrom)
                : null;
        this.id = payload.id;
        this.createdAt = deserialize.iso8601DateTime(payload.createdAt);
        this.updatedAt = deserialize.iso8601DateTime(payload.updatedAt);
        this.version = deserialize.integer(payload.version);
        this._solution = { storeId, dataMappingId: dataMappingId };
    }
    get _proxy() {
        this._context =
            this._context ||
                new DataMappingContextImpl(this._version, this._solution.storeId, this._solution.dataMappingId);
        return this._context;
    }
    /**
     * Remove a DataMappingInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed DataMappingInstance
     */
    remove(callback) {
        return this._proxy.remove(callback);
    }
    /**
     * Remove a DataMappingInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed DataMappingInstance with HTTP metadata
     */
    removeWithHttpInfo(callback) {
        return this._proxy.removeWithHttpInfo(callback);
    }
    /**
     * Fetch a DataMappingInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed DataMappingInstance
     */
    fetch(callback) {
        return this._proxy.fetch(callback);
    }
    /**
     * Fetch a DataMappingInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed DataMappingInstance with HTTP metadata
     */
    fetchWithHttpInfo(callback) {
        return this._proxy.fetchWithHttpInfo(callback);
    }
    patch(params, callback) {
        return this._proxy.patch(params, callback);
    }
    patchWithHttpInfo(params, callback) {
        return this._proxy.patchWithHttpInfo(params, callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            message: this.message,
            statusUrl: this.statusUrl,
            displayName: this.displayName,
            description: this.description,
            isEnabled: this.isEnabled,
            mappingTo: this.mappingTo,
            mappingFrom: this.mappingFrom,
            id: this.id,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
            version: this.version,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.DataMappingInstance = DataMappingInstance;
function DataMappingListInstance(version, storeId) {
    if (!(0, utility_1.isValidPathParam)(storeId)) {
        throw new Error("Parameter 'storeId' is not valid.");
    }
    const instance = ((dataMappingId) => instance.get(dataMappingId));
    instance.get = function get(dataMappingId) {
        return new DataMappingContextImpl(version, storeId, dataMappingId);
    };
    instance._version = version;
    instance._solution = { storeId };
    instance._uri = `/ControlPlane/Stores/${storeId}/DataMappings`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new DataMappingInstance(operationVersion, payload, instance._solution.storeId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new DataMappingInstance(operationVersion, response.body, instance._solution.storeId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.page = function page(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        if (params["type"] !== undefined)
            data["type"] = params["type"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.page({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new DataMappingPage(operationVersion, payload, instance._uri, data, instance._solution));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.list = instance._version.list;
    instance.getPage = function getPage(targetUrl, callback) {
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((payload) => new DataMappingPage(instance._version, payload, instance._uri, {}, instance._solution));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.pageWithHttpInfo = function pageWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        if (params["type"] !== undefined)
            data["type"] = params["type"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // For page operations, use page() directly as it already returns { statusCode, body, headers }
        // IMPORTANT: Pass full response to Page constructor, not response.body
        let operationPromise = operationVersion
            .page({ uri: instance._uri, method: "get", params: data, headers })
            .then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new DataMappingPage(operationVersion, response, instance._uri, data, instance._solution),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.eachWithHttpInfo = instance._version.eachWithHttpInfo;
    instance.list = instance._version.list;
    instance.listWithHttpInfo = instance._version.listWithHttpInfo;
    instance.getPageWithHttpInfo = function getPageWithHttpInfo(targetUrl, callback) {
        // Use request() directly as it already returns { statusCode, body, headers }
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new DataMappingPage(instance._version, response, instance._uri, {}, instance._solution),
        }));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
class DataMappingPage extends TokenPage_1.default {
    /**
     * Initialize the DataMappingPage
     *
     * @param version - Version of the resource
     * @param response - Response from the API
     * @param uri - URI of the resource
     * @param params - Query parameters
     * @param solution - Path solution
     */
    constructor(version, response, uri, params, solution) {
        super(version, response, uri, params, solution);
    }
    /**
     * Build an instance of DataMappingInstance
     *
     * @param payload - Payload response from the API
     */
    getInstance(payload) {
        return new DataMappingInstance(this._version, payload, this._solution.storeId);
    }
    [util_1.inspect.custom](depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.DataMappingPage = DataMappingPage;
