// twilio/lib/rest/memory/v1/import.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Memory API
 * APIs for managing memory stores, profiles, events, and conversational intelligence capabilities.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportInstance = exports.ImportContextImpl = exports.FetchProfileImportV2200ResponseSummary = exports.CreateProfilesImportV2Request = exports.ColumnMappingItem = void 0;
exports.ImportListInstance = ImportListInstance;
const util_1 = require("util");
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
/**
 * Mapping of a CSV header column to the trait fields
 */
class ColumnMappingItem {
    constructor(payload) {
        this.columnName = payload["columnName"];
        this.traitGroup = payload["traitGroup"];
        this.traitName = payload["traitName"];
    }
}
exports.ColumnMappingItem = ColumnMappingItem;
class CreateProfilesImportV2Request {
    constructor(payload) {
        this.filename = payload["filename"];
        this.fileSize = payload["fileSize"];
        this.columnMappings = payload["columnMappings"];
    }
}
exports.CreateProfilesImportV2Request = CreateProfilesImportV2Request;
/**
 * Summary statistics of the import operation
 */
class FetchProfileImportV2200ResponseSummary {
    constructor(payload) {
        this.errors = payload["errors"];
        this.warnings = payload["warnings"];
    }
}
exports.FetchProfileImportV2200ResponseSummary = FetchProfileImportV2200ResponseSummary;
class ImportContextImpl {
    constructor(_version, storeId, importId) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(storeId)) {
            throw new Error("Parameter 'storeId' is not valid.");
        }
        if (!(0, utility_1.isValidPathParam)(importId)) {
            throw new Error("Parameter 'importId' is not valid.");
        }
        this._solution = { storeId, importId };
        this._uri = `/Stores/${storeId}/Profiles/Imports/${importId}`;
    }
    fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ImportInstance(operationVersion, payload, instance._solution.storeId, instance._solution.importId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ImportInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.importId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ImportContextImpl = ImportContextImpl;
class ImportInstance {
    constructor(_version, _payload, storeId, importId) {
        this._version = _version;
        const payload = _payload;
        this.status = payload.status;
        this.filename = payload.filename;
        this.createdAt = deserialize.iso8601DateTime(payload.createdAt);
        this.updatedAt = deserialize.iso8601DateTime(payload.updatedAt);
        this.fileSize = deserialize.integer(payload.fileSize);
        this.columnMappings =
            payload.columnMappings !== null && payload.columnMappings !== undefined
                ? payload.columnMappings.map((payload) => new ColumnMappingItem(payload))
                : null;
        this.summary =
            payload.summary !== null && payload.summary !== undefined
                ? new FetchProfileImportV2200ResponseSummary(payload.summary)
                : null;
        this.importId = payload.importId;
        this.url = payload.url;
        this.imports = payload.imports;
        this._solution = { storeId, importId: importId };
    }
    get _proxy() {
        this._context =
            this._context ||
                new ImportContextImpl(this._version, this._solution.storeId, this._solution.importId);
        return this._context;
    }
    /**
     * Fetch a ImportInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ImportInstance
     */
    fetch(callback) {
        return this._proxy.fetch(callback);
    }
    /**
     * Fetch a ImportInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ImportInstance with HTTP metadata
     */
    fetchWithHttpInfo(callback) {
        return this._proxy.fetchWithHttpInfo(callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            status: this.status,
            filename: this.filename,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
            fileSize: this.fileSize,
            columnMappings: this.columnMappings,
            summary: this.summary,
            importId: this.importId,
            url: this.url,
            imports: this.imports,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ImportInstance = ImportInstance;
function ImportListInstance(version, storeId) {
    if (!(0, utility_1.isValidPathParam)(storeId)) {
        throw new Error("Parameter 'storeId' is not valid.");
    }
    const instance = ((importId) => instance.get(importId));
    instance.get = function get(importId) {
        return new ImportContextImpl(version, storeId, importId);
    };
    instance._version = version;
    instance._solution = { storeId };
    instance._uri = `/Stores/${storeId}/Profiles/Imports`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ImportInstance(operationVersion, payload, instance._solution.storeId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ImportInstance(operationVersion, response.body, instance._solution.storeId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.list = function list(callback) {
        const headers = {};
        const data = {};
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        // Array items are primitives (strings, numbers), return them directly
        operationPromise = operationPromise.then((payload) => payload["imports"] || []);
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.listWithHttpInfo = function listWithHttpInfo(callback) {
        const headers = {};
        const data = {};
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // Array items are primitives (strings, numbers), return them directly
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: response.body["imports"] || [],
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
