// twilio/lib/rest/memory/v1/observation.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Memory API
 * APIs for managing memory stores, profiles, events, and conversational intelligence capabilities.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObservationPage = exports.ObservationInstance = exports.ObservationContextImpl = exports.ObservationCore = exports.CreateObservationsRequest = void 0;
exports.ObservationListInstance = ObservationListInstance;
const util_1 = require("util");
const TokenPage_1 = __importDefault(require("../../../base/TokenPage"));
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
/**
 * Request payload for creating one or more observations. Supports creation of up to 10 observations.
 */
class CreateObservationsRequest {
    constructor(payload) {
        this.observations = payload["observations"];
    }
}
exports.CreateObservationsRequest = CreateObservationsRequest;
/**
 * Core mutable properties for creating observation objects.
 */
class ObservationCore {
    constructor(payload) {
        this.content = payload["content"];
        this.occurredAt = payload["occurredAt"];
        this.source = payload["source"];
        this.conversationIds = payload["conversationIds"];
    }
}
exports.ObservationCore = ObservationCore;
class ObservationContextImpl {
    constructor(_version, storeId, profileId, observationId) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(storeId)) {
            throw new Error("Parameter 'storeId' is not valid.");
        }
        if (!(0, utility_1.isValidPathParam)(profileId)) {
            throw new Error("Parameter 'profileId' is not valid.");
        }
        if (!(0, utility_1.isValidPathParam)(observationId)) {
            throw new Error("Parameter 'observationId' is not valid.");
        }
        this._solution = { storeId, profileId, observationId };
        this._uri = `/Stores/${storeId}/Profiles/${profileId}/Observations/${observationId}`;
    }
    remove(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "delete",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ObservationInstance(operationVersion, payload, instance._solution.storeId, instance._solution.profileId, instance._solution.observationId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    removeWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // DELETE operation that returns a response model
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "delete",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ObservationInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.profileId, instance._solution.observationId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ObservationInstance(operationVersion, payload, instance._solution.storeId, instance._solution.profileId, instance._solution.observationId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ObservationInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.profileId, instance._solution.observationId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patch(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.patch({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ObservationInstance(operationVersion, payload, instance._solution.storeId, instance._solution.profileId, instance._solution.observationId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patchWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .patchWithResponseInfo({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ObservationInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.profileId, instance._solution.observationId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ObservationContextImpl = ObservationContextImpl;
/**
 * Response for batch observation creation.
 */
class ObservationInstance {
    constructor(_version, _payload, storeId, profileId, observationId) {
        this._version = _version;
        const payload = _payload;
        this.message = payload.message;
        this.content = payload.content;
        this.occurredAt = deserialize.iso8601DateTime(payload.occurredAt);
        this.source = payload.source;
        this.conversationIds = payload.conversationIds;
        this.id = payload.id;
        this.createdAt = deserialize.iso8601DateTime(payload.createdAt);
        this.updatedAt = deserialize.iso8601DateTime(payload.updatedAt);
        this._solution = { storeId, profileId, observationId: observationId };
    }
    get _proxy() {
        this._context =
            this._context ||
                new ObservationContextImpl(this._version, this._solution.storeId, this._solution.profileId, this._solution.observationId);
        return this._context;
    }
    /**
     * Remove a ObservationInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ObservationInstance
     */
    remove(callback) {
        return this._proxy.remove(callback);
    }
    /**
     * Remove a ObservationInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ObservationInstance with HTTP metadata
     */
    removeWithHttpInfo(callback) {
        return this._proxy.removeWithHttpInfo(callback);
    }
    /**
     * Fetch a ObservationInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ObservationInstance
     */
    fetch(callback) {
        return this._proxy.fetch(callback);
    }
    /**
     * Fetch a ObservationInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ObservationInstance with HTTP metadata
     */
    fetchWithHttpInfo(callback) {
        return this._proxy.fetchWithHttpInfo(callback);
    }
    patch(params, callback) {
        return this._proxy.patch(params, callback);
    }
    patchWithHttpInfo(params, callback) {
        return this._proxy.patchWithHttpInfo(params, callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            message: this.message,
            content: this.content,
            occurredAt: this.occurredAt,
            source: this.source,
            conversationIds: this.conversationIds,
            id: this.id,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ObservationInstance = ObservationInstance;
function ObservationListInstance(version, storeId, profileId) {
    if (!(0, utility_1.isValidPathParam)(storeId)) {
        throw new Error("Parameter 'storeId' is not valid.");
    }
    if (!(0, utility_1.isValidPathParam)(profileId)) {
        throw new Error("Parameter 'profileId' is not valid.");
    }
    const instance = ((observationId) => instance.get(observationId));
    instance.get = function get(observationId) {
        return new ObservationContextImpl(version, storeId, profileId, observationId);
    };
    instance._version = version;
    instance._solution = { storeId, profileId };
    instance._uri = `/Stores/${storeId}/Profiles/${profileId}/Observations`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ObservationInstance(operationVersion, payload, instance._solution.storeId, instance._solution.profileId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ObservationInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.profileId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.page = function page(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        if (params["source"] !== undefined)
            data["source"] = params["source"];
        if (params["createdAfter"] !== undefined)
            data["createdAfter"] = serialize.iso8601DateTime(params["createdAfter"]);
        if (params["createdBefore"] !== undefined)
            data["createdBefore"] = serialize.iso8601DateTime(params["createdBefore"]);
        const headers = {};
        headers["Accept"] = "application/json";
        if (params["acceptEncoding"] !== undefined)
            headers["Accept-Encoding"] = params["acceptEncoding"];
        let operationVersion = version, operationPromise = operationVersion.page({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ObservationPage(operationVersion, payload, instance._uri, data, instance._solution));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.list = instance._version.list;
    instance.getPage = function getPage(targetUrl, callback) {
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((payload) => new ObservationPage(instance._version, payload, instance._uri, {}, instance._solution));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.pageWithHttpInfo = function pageWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        if (params["source"] !== undefined)
            data["source"] = params["source"];
        if (params["createdAfter"] !== undefined)
            data["createdAfter"] = serialize.iso8601DateTime(params["createdAfter"]);
        if (params["createdBefore"] !== undefined)
            data["createdBefore"] = serialize.iso8601DateTime(params["createdBefore"]);
        const headers = {};
        headers["Accept"] = "application/json";
        if (params["acceptEncoding"] !== undefined)
            headers["Accept-Encoding"] = params["acceptEncoding"];
        let operationVersion = version;
        // For page operations, use page() directly as it already returns { statusCode, body, headers }
        // IMPORTANT: Pass full response to Page constructor, not response.body
        let operationPromise = operationVersion
            .page({ uri: instance._uri, method: "get", params: data, headers })
            .then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new ObservationPage(operationVersion, response, instance._uri, data, instance._solution),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.eachWithHttpInfo = instance._version.eachWithHttpInfo;
    instance.list = instance._version.list;
    instance.listWithHttpInfo = instance._version.listWithHttpInfo;
    instance.getPageWithHttpInfo = function getPageWithHttpInfo(targetUrl, callback) {
        // Use request() directly as it already returns { statusCode, body, headers }
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new ObservationPage(instance._version, response, instance._uri, {}, instance._solution),
        }));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
class ObservationPage extends TokenPage_1.default {
    /**
     * Initialize the ObservationPage
     *
     * @param version - Version of the resource
     * @param response - Response from the API
     * @param uri - URI of the resource
     * @param params - Query parameters
     * @param solution - Path solution
     */
    constructor(version, response, uri, params, solution) {
        super(version, response, uri, params, solution);
    }
    /**
     * Build an instance of ObservationInstance
     *
     * @param payload - Payload response from the API
     */
    getInstance(payload) {
        return new ObservationInstance(this._version, payload, this._solution.storeId, this._solution.profileId);
    }
    [util_1.inspect.custom](depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ObservationPage = ObservationPage;
