// twilio/lib/rest/memory/v1/profile.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Memory API
 * APIs for managing memory stores, profiles, events, and conversational intelligence capabilities.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProfilePage = exports.ProfileInstance = exports.ProfileContextImpl = exports.ProfilesMeta = exports.ProfilePatch = exports.ProfileData = void 0;
exports.ProfileListInstance = ProfileListInstance;
const util_1 = require("util");
const TokenPage_1 = __importDefault(require("../../../base/TokenPage"));
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
/**
 * Data for creating or updating a profile, including traits.
 */
class ProfileData {
    constructor(payload) {
        this.traits = payload["traits"];
    }
}
exports.ProfileData = ProfileData;
class ProfilePatch {
    constructor(payload) {
        this.traits = payload["traits"];
    }
}
exports.ProfilePatch = ProfilePatch;
class ProfilesMeta {
    constructor(payload) {
        this.key = payload["key"];
        this.pageSize = payload["pageSize"];
        this.nextToken = payload["nextToken"];
        this.previousToken = payload["previousToken"];
    }
}
exports.ProfilesMeta = ProfilesMeta;
class ProfileContextImpl {
    constructor(_version, storeId, profileId) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(storeId)) {
            throw new Error("Parameter 'storeId' is not valid.");
        }
        if (!(0, utility_1.isValidPathParam)(profileId)) {
            throw new Error("Parameter 'profileId' is not valid.");
        }
        this._solution = { storeId, profileId };
        this._uri = `/Stores/${storeId}/Profiles/${profileId}`;
    }
    remove(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "delete",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ProfileInstance(operationVersion, payload, instance._solution.storeId, instance._solution.profileId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    removeWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // DELETE operation that returns a response model
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "delete",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ProfileInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.profileId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetch(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["traitGroups"] !== undefined)
            data["traitGroups"] = params["traitGroups"];
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ProfileInstance(operationVersion, payload, instance._solution.storeId, instance._solution.profileId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["traitGroups"] !== undefined)
            data["traitGroups"] = params["traitGroups"];
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ProfileInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.profileId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patch(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.patch({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ProfileInstance(operationVersion, payload, instance._solution.storeId, instance._solution.profileId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patchWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .patchWithResponseInfo({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ProfileInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.profileId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ProfileContextImpl = ProfileContextImpl;
class ProfileInstance {
    constructor(_version, _payload, storeId, profileId) {
        this._version = _version;
        const payload = _payload;
        this.id = payload.id;
        this.message = payload.message;
        this.profiles = payload.profiles;
        this.meta =
            payload.meta !== null && payload.meta !== undefined
                ? new ProfilesMeta(payload.meta)
                : null;
        this.createdAt = deserialize.iso8601DateTime(payload.createdAt);
        this.traits = payload.traits;
        this._solution = { storeId, profileId: profileId };
    }
    get _proxy() {
        this._context =
            this._context ||
                new ProfileContextImpl(this._version, this._solution.storeId, this._solution.profileId);
        return this._context;
    }
    /**
     * Remove a ProfileInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ProfileInstance
     */
    remove(callback) {
        return this._proxy.remove(callback);
    }
    /**
     * Remove a ProfileInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ProfileInstance with HTTP metadata
     */
    removeWithHttpInfo(callback) {
        return this._proxy.removeWithHttpInfo(callback);
    }
    fetch(params, callback) {
        return this._proxy.fetch(params, callback);
    }
    fetchWithHttpInfo(params, callback) {
        return this._proxy.fetchWithHttpInfo(params, callback);
    }
    patch(params, callback) {
        return this._proxy.patch(params, callback);
    }
    patchWithHttpInfo(params, callback) {
        return this._proxy.patchWithHttpInfo(params, callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            id: this.id,
            message: this.message,
            profiles: this.profiles,
            meta: this.meta,
            createdAt: this.createdAt,
            traits: this.traits,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ProfileInstance = ProfileInstance;
function ProfileListInstance(version, storeId) {
    if (!(0, utility_1.isValidPathParam)(storeId)) {
        throw new Error("Parameter 'storeId' is not valid.");
    }
    const instance = ((profileId) => instance.get(profileId));
    instance.get = function get(profileId) {
        return new ProfileContextImpl(version, storeId, profileId);
    };
    instance._version = version;
    instance._solution = { storeId };
    instance._uri = `/Stores/${storeId}/Profiles`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ProfileInstance(operationVersion, payload, instance._solution.storeId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ProfileInstance(operationVersion, response.body, instance._solution.storeId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.page = function page(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.page({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ProfilePage(operationVersion, payload, instance._uri, data, instance._solution));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.list = function list(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        params = params || {};
        let allRecords = [];
        function fetchNextPage(nextPageParams) {
            return instance.page(nextPageParams).then((page) => {
                const payload = page["_payload"];
                const records = payload && payload["profiles"] ? payload["profiles"] : [];
                allRecords = allRecords.concat(records);
                if (params.limit && allRecords.length >= params.limit) {
                    allRecords = allRecords.slice(0, params.limit);
                    return allRecords;
                }
                const nextToken = page.nextPageUrl
                    ? new URL(page.nextPageUrl).searchParams.get("PageToken")
                    : null;
                if (nextToken) {
                    return fetchNextPage({ ...nextPageParams, pageToken: nextToken });
                }
                return allRecords;
            });
        }
        const promise = fetchNextPage(params);
        return instance._version.setPromiseCallback(promise, callback);
    };
    instance.getPage = function getPage(targetUrl, callback) {
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((payload) => new ProfilePage(instance._version, payload, instance._uri, {}, instance._solution));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.pageWithHttpInfo = function pageWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // For page operations, use page() directly as it already returns { statusCode, body, headers }
        // IMPORTANT: Pass full response to Page constructor, not response.body
        let operationPromise = operationVersion
            .page({ uri: instance._uri, method: "get", params: data, headers })
            .then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new ProfilePage(operationVersion, response, instance._uri, data, instance._solution),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.eachWithHttpInfo = instance._version.eachWithHttpInfo;
    instance.listWithHttpInfo = function listWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        params = params || {};
        let allRecords = [];
        let lastResponse;
        function fetchNextPage(nextPageParams) {
            return instance
                .pageWithHttpInfo(nextPageParams)
                .then((response) => {
                lastResponse = response;
                const page = response.body;
                // For primitive arrays, directly access the records array from the payload
                // Access _payload using bracket notation to bypass TypeScript protected access
                const payload = page["_payload"];
                const records = payload && payload["profiles"] ? payload["profiles"] : [];
                allRecords = allRecords.concat(records);
                if (params.limit && allRecords.length >= params.limit) {
                    allRecords = allRecords.slice(0, params.limit);
                    return {
                        statusCode: lastResponse.statusCode,
                        headers: lastResponse.headers,
                        body: allRecords,
                    };
                }
                const nextToken = page.nextPageUrl
                    ? new URL(page.nextPageUrl).searchParams.get("PageToken")
                    : null;
                if (nextToken) {
                    return fetchNextPage({ ...nextPageParams, pageToken: nextToken });
                }
                return {
                    statusCode: lastResponse.statusCode,
                    headers: lastResponse.headers,
                    body: allRecords,
                };
            });
        }
        const promise = fetchNextPage(params);
        return instance._version.setPromiseCallback(promise, callback);
    };
    instance.getPageWithHttpInfo = function getPageWithHttpInfo(targetUrl, callback) {
        // Use request() directly as it already returns { statusCode, body, headers }
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new ProfilePage(instance._version, response, instance._uri, {}, instance._solution),
        }));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
class ProfilePage extends TokenPage_1.default {
    /**
     * Initialize the ProfilePage
     *
     * @param version - Version of the resource
     * @param response - Response from the API
     * @param uri - URI of the resource
     * @param params - Query parameters
     * @param solution - Path solution
     */
    constructor(version, response, uri, params, solution) {
        super(version, response, uri, params, solution);
    }
    /**
     * Build an instance of ProfileInstance
     *
     * @param payload - Payload response from the API
     */
    getInstance(payload) {
        // Array items are primitives (strings, numbers), return directly
        return payload;
    }
    [util_1.inspect.custom](depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ProfilePage = ProfilePage;
