// twilio/lib/rest/memory/v1/store.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Memory API
 * APIs for managing memory stores, profiles, events, and conversational intelligence capabilities.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorePage = exports.StoreInstance = exports.StoreContextImpl = exports.ServiceRequest = exports.PatchStoreRequest = exports.Meta = void 0;
exports.StoreListInstance = StoreListInstance;
const util_1 = require("util");
const TokenPage_1 = __importDefault(require("../../../base/TokenPage"));
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
class Meta {
    constructor(payload) {
        this.key = payload["key"];
        this.pageSize = payload["pageSize"];
        this.nextToken = payload["nextToken"];
        this.previousToken = payload["previousToken"];
    }
}
exports.Meta = Meta;
class PatchStoreRequest {
    constructor(payload) {
        this.displayName = payload["displayName"];
        this.description = payload["description"];
    }
}
exports.PatchStoreRequest = PatchStoreRequest;
class ServiceRequest {
    constructor(payload) {
        this.displayName = payload["displayName"];
        this.description = payload["description"];
    }
}
exports.ServiceRequest = ServiceRequest;
class StoreContextImpl {
    constructor(_version, storeId) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(storeId)) {
            throw new Error("Parameter 'storeId' is not valid.");
        }
        this._solution = { storeId };
        this._uri = `/ControlPlane/Stores/${storeId}`;
    }
    remove(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "delete",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new StoreInstance(operationVersion, payload, instance._solution.storeId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    removeWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // DELETE operation that returns a response model
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "delete",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new StoreInstance(operationVersion, response.body, instance._solution.storeId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new StoreInstance(operationVersion, payload, instance._solution.storeId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new StoreInstance(operationVersion, response.body, instance._solution.storeId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patch(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params =
                params || {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.patch({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new StoreInstance(operationVersion, payload, instance._solution.storeId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patchWithHttpInfo(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params =
                params || {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .patchWithResponseInfo({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new StoreInstance(operationVersion, response.body, instance._solution.storeId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.StoreContextImpl = StoreContextImpl;
class StoreInstance {
    constructor(_version, _payload, storeId) {
        this._version = _version;
        const payload = _payload;
        this.message = payload.message;
        this.statusUrl = payload.statusUrl;
        this.stores = payload.stores;
        this.meta =
            payload.meta !== null && payload.meta !== undefined
                ? new Meta(payload.meta)
                : null;
        this.displayName = payload.displayName;
        this.description = payload.description;
        this.id = payload.id;
        this.status = payload.status;
        this.intelligenceServiceId = payload.intelligenceServiceId;
        this.version = deserialize.integer(payload.version);
        this._solution = { storeId: storeId };
    }
    get _proxy() {
        this._context =
            this._context ||
                new StoreContextImpl(this._version, this._solution.storeId);
        return this._context;
    }
    /**
     * Remove a StoreInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed StoreInstance
     */
    remove(callback) {
        return this._proxy.remove(callback);
    }
    /**
     * Remove a StoreInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed StoreInstance with HTTP metadata
     */
    removeWithHttpInfo(callback) {
        return this._proxy.removeWithHttpInfo(callback);
    }
    /**
     * Fetch a StoreInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed StoreInstance
     */
    fetch(callback) {
        return this._proxy.fetch(callback);
    }
    /**
     * Fetch a StoreInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed StoreInstance with HTTP metadata
     */
    fetchWithHttpInfo(callback) {
        return this._proxy.fetchWithHttpInfo(callback);
    }
    patch(params, callback) {
        return this._proxy.patch(params, callback);
    }
    patchWithHttpInfo(params, callback) {
        return this._proxy.patchWithHttpInfo(params, callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            message: this.message,
            statusUrl: this.statusUrl,
            stores: this.stores,
            meta: this.meta,
            displayName: this.displayName,
            description: this.description,
            id: this.id,
            status: this.status,
            intelligenceServiceId: this.intelligenceServiceId,
            version: this.version,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.StoreInstance = StoreInstance;
function StoreListInstance(version) {
    const instance = ((storeId) => instance.get(storeId));
    instance.get = function get(storeId) {
        return new StoreContextImpl(version, storeId);
    };
    instance._version = version;
    instance._solution = {};
    instance._uri = `/ControlPlane/Stores`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new StoreInstance(operationVersion, payload));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new StoreInstance(operationVersion, response.body),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.page = function page(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.page({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new StorePage(operationVersion, payload, instance._uri, data, instance._solution));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.list = function list(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        params = params || {};
        let allRecords = [];
        function fetchNextPage(nextPageParams) {
            return instance.page(nextPageParams).then((page) => {
                const payload = page["_payload"];
                const records = payload && payload["stores"] ? payload["stores"] : [];
                allRecords = allRecords.concat(records);
                if (params.limit && allRecords.length >= params.limit) {
                    allRecords = allRecords.slice(0, params.limit);
                    return allRecords;
                }
                const nextToken = page.nextPageUrl
                    ? new URL(page.nextPageUrl).searchParams.get("PageToken")
                    : null;
                if (nextToken) {
                    return fetchNextPage({ ...nextPageParams, pageToken: nextToken });
                }
                return allRecords;
            });
        }
        const promise = fetchNextPage(params);
        return instance._version.setPromiseCallback(promise, callback);
    };
    instance.getPage = function getPage(targetUrl, callback) {
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((payload) => new StorePage(instance._version, payload, instance._uri, {}, instance._solution));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.pageWithHttpInfo = function pageWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // For page operations, use page() directly as it already returns { statusCode, body, headers }
        // IMPORTANT: Pass full response to Page constructor, not response.body
        let operationPromise = operationVersion
            .page({ uri: instance._uri, method: "get", params: data, headers })
            .then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new StorePage(operationVersion, response, instance._uri, data, instance._solution),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.eachWithHttpInfo = instance._version.eachWithHttpInfo;
    instance.listWithHttpInfo = function listWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        params = params || {};
        let allRecords = [];
        let lastResponse;
        function fetchNextPage(nextPageParams) {
            return instance
                .pageWithHttpInfo(nextPageParams)
                .then((response) => {
                lastResponse = response;
                const page = response.body;
                // For primitive arrays, directly access the records array from the payload
                // Access _payload using bracket notation to bypass TypeScript protected access
                const payload = page["_payload"];
                const records = payload && payload["stores"] ? payload["stores"] : [];
                allRecords = allRecords.concat(records);
                if (params.limit && allRecords.length >= params.limit) {
                    allRecords = allRecords.slice(0, params.limit);
                    return {
                        statusCode: lastResponse.statusCode,
                        headers: lastResponse.headers,
                        body: allRecords,
                    };
                }
                const nextToken = page.nextPageUrl
                    ? new URL(page.nextPageUrl).searchParams.get("PageToken")
                    : null;
                if (nextToken) {
                    return fetchNextPage({ ...nextPageParams, pageToken: nextToken });
                }
                return {
                    statusCode: lastResponse.statusCode,
                    headers: lastResponse.headers,
                    body: allRecords,
                };
            });
        }
        const promise = fetchNextPage(params);
        return instance._version.setPromiseCallback(promise, callback);
    };
    instance.getPageWithHttpInfo = function getPageWithHttpInfo(targetUrl, callback) {
        // Use request() directly as it already returns { statusCode, body, headers }
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new StorePage(instance._version, response, instance._uri, {}, instance._solution),
        }));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
class StorePage extends TokenPage_1.default {
    /**
     * Initialize the StorePage
     *
     * @param version - Version of the resource
     * @param response - Response from the API
     * @param uri - URI of the resource
     * @param params - Query parameters
     * @param solution - Path solution
     */
    constructor(version, response, uri, params, solution) {
        super(version, response, uri, params, solution);
    }
    /**
     * Build an instance of StoreInstance
     *
     * @param payload - Payload response from the API
     */
    getInstance(payload) {
        // Array items are primitives (strings, numbers), return directly
        return payload;
    }
    [util_1.inspect.custom](depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.StorePage = StorePage;
