// twilio/lib/rest/memory/v1/traitGroup.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Memory API
 * APIs for managing memory stores, profiles, events, and conversational intelligence capabilities.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraitGroupPage = exports.TraitGroupInstance = exports.TraitGroupContextImpl = exports.ValidationRule = exports.TraitGroupRequest = exports.TraitGroupCoreTraitsValue = exports.TraitGroup = exports.TraitDefinition = exports.PatchTraitGroupRequest = exports.Meta = void 0;
exports.TraitGroupListInstance = TraitGroupListInstance;
const util_1 = require("util");
const TokenPage_1 = __importDefault(require("../../../base/TokenPage"));
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
class Meta {
    constructor(payload) {
        this.key = payload["key"];
        this.pageSize = payload["pageSize"];
        this.nextToken = payload["nextToken"];
        this.previousToken = payload["previousToken"];
    }
}
exports.Meta = Meta;
class PatchTraitGroupRequest {
    constructor(payload) {
        this.description = payload["description"];
        this.traits = payload["traits"];
    }
}
exports.PatchTraitGroupRequest = PatchTraitGroupRequest;
/**
 * A Trait is a single attribute or characteristic of a profile, such as age, location, or preferences.
 */
class TraitDefinition {
    constructor(payload) {
        this.dataType = payload["dataType"];
        this.description = payload["description"];
        this.idTypePromotion = payload["idTypePromotion"];
    }
}
exports.TraitDefinition = TraitDefinition;
class TraitGroup {
    constructor(payload) {
        this.displayName = payload["displayName"];
        this.description = payload["description"];
        this.traits = payload["traits"];
        this.version = payload["version"];
    }
}
exports.TraitGroup = TraitGroup;
/**
 * A Trait is a single attribute or characteristic of a profile, such as age, location, or preferences.
 */
class TraitGroupCoreTraitsValue {
    constructor(payload) {
        this.dataType = payload["dataType"];
        this.description = payload["description"];
        this.validationRule = payload["validationRule"];
        this.idTypePromotion = payload["idTypePromotion"];
    }
}
exports.TraitGroupCoreTraitsValue = TraitGroupCoreTraitsValue;
/**
 * Payload schema for creating Trait Groups.
 */
class TraitGroupRequest {
    constructor(payload) {
        this.displayName = payload["displayName"];
        this.description = payload["description"];
        this.traits = payload["traits"];
    }
}
exports.TraitGroupRequest = TraitGroupRequest;
class ValidationRule {
    constructor(payload) {
        this.pattern = payload["pattern"];
        this.minLength = payload["minLength"];
        this.maxLength = payload["maxLength"];
        this.ruleType = payload["ruleType"];
        this.min = payload["min"];
        this.max = payload["max"];
        this.minItems = payload["minItems"];
        this.maxItems = payload["maxItems"];
    }
}
exports.ValidationRule = ValidationRule;
class TraitGroupContextImpl {
    constructor(_version, storeId, traitGroupName) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(storeId)) {
            throw new Error("Parameter 'storeId' is not valid.");
        }
        if (!(0, utility_1.isValidPathParam)(traitGroupName)) {
            throw new Error("Parameter 'traitGroupName' is not valid.");
        }
        this._solution = { storeId, traitGroupName };
        this._uri = `/ControlPlane/Stores/${storeId}/TraitGroups/${traitGroupName}`;
    }
    remove(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "delete",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new TraitGroupInstance(operationVersion, payload, instance._solution.storeId, instance._solution.traitGroupName));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    removeWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // DELETE operation that returns a response model
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "delete",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new TraitGroupInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.traitGroupName),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetch(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["includeTraits"] !== undefined)
            data["includeTraits"] = serialize.bool(params["includeTraits"]);
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new TraitGroupInstance(operationVersion, payload, instance._solution.storeId, instance._solution.traitGroupName));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["includeTraits"] !== undefined)
            data["includeTraits"] = serialize.bool(params["includeTraits"]);
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new TraitGroupInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.traitGroupName),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patch(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params =
                params ||
                    {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.patch({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new TraitGroupInstance(operationVersion, payload, instance._solution.storeId, instance._solution.traitGroupName));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    patchWithHttpInfo(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params =
                params ||
                    {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .patchWithResponseInfo({
            uri: instance._uri,
            method: "patch",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new TraitGroupInstance(operationVersion, response.body, instance._solution.storeId, instance._solution.traitGroupName),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.TraitGroupContextImpl = TraitGroupContextImpl;
class TraitGroupInstance {
    constructor(_version, _payload, storeId, traitGroupName) {
        this._version = _version;
        const payload = _payload;
        this.message = payload.message;
        this.statusUrl = payload.statusUrl;
        this.displayName = payload.displayName;
        this.description = payload.description;
        this.traits = payload.traits;
        this.version = deserialize.integer(payload.version);
        this.traitGroup =
            payload.traitGroup !== null && payload.traitGroup !== undefined
                ? new TraitGroup(payload.traitGroup)
                : null;
        this.meta =
            payload.meta !== null && payload.meta !== undefined
                ? new Meta(payload.meta)
                : null;
        this._solution = { storeId, traitGroupName: traitGroupName };
    }
    get _proxy() {
        this._context =
            this._context ||
                new TraitGroupContextImpl(this._version, this._solution.storeId, this._solution.traitGroupName);
        return this._context;
    }
    /**
     * Remove a TraitGroupInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed TraitGroupInstance
     */
    remove(callback) {
        return this._proxy.remove(callback);
    }
    /**
     * Remove a TraitGroupInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed TraitGroupInstance with HTTP metadata
     */
    removeWithHttpInfo(callback) {
        return this._proxy.removeWithHttpInfo(callback);
    }
    fetch(params, callback) {
        return this._proxy.fetch(params, callback);
    }
    fetchWithHttpInfo(params, callback) {
        return this._proxy.fetchWithHttpInfo(params, callback);
    }
    patch(params, callback) {
        return this._proxy.patch(params, callback);
    }
    patchWithHttpInfo(params, callback) {
        return this._proxy.patchWithHttpInfo(params, callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            message: this.message,
            statusUrl: this.statusUrl,
            displayName: this.displayName,
            description: this.description,
            traits: this.traits,
            version: this.version,
            traitGroup: this.traitGroup,
            meta: this.meta,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.TraitGroupInstance = TraitGroupInstance;
function TraitGroupListInstance(version, storeId) {
    if (!(0, utility_1.isValidPathParam)(storeId)) {
        throw new Error("Parameter 'storeId' is not valid.");
    }
    const instance = ((traitGroupName) => instance.get(traitGroupName));
    instance.get = function get(traitGroupName) {
        return new TraitGroupContextImpl(version, storeId, traitGroupName);
    };
    instance._version = version;
    instance._solution = { storeId };
    instance._uri = `/ControlPlane/Stores/${storeId}/TraitGroups`;
    instance.create = function create(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params =
                params || {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new TraitGroupInstance(operationVersion, payload, instance._solution.storeId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params =
                params || {};
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new TraitGroupInstance(operationVersion, response.body, instance._solution.storeId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.page = function page(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["includeTraits"] !== undefined)
            data["includeTraits"] = serialize.bool(params["includeTraits"]);
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.page({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new TraitGroupPage(operationVersion, payload, instance._uri, data, instance._solution));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.list = instance._version.list;
    instance.getPage = function getPage(targetUrl, callback) {
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((payload) => new TraitGroupPage(instance._version, payload, instance._uri, {}, instance._solution));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.pageWithHttpInfo = function pageWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["includeTraits"] !== undefined)
            data["includeTraits"] = serialize.bool(params["includeTraits"]);
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["orderBy"] !== undefined)
            data["orderBy"] = params["orderBy"];
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // For page operations, use page() directly as it already returns { statusCode, body, headers }
        // IMPORTANT: Pass full response to Page constructor, not response.body
        let operationPromise = operationVersion
            .page({ uri: instance._uri, method: "get", params: data, headers })
            .then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new TraitGroupPage(operationVersion, response, instance._uri, data, instance._solution),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.eachWithHttpInfo = instance._version.eachWithHttpInfo;
    instance.list = instance._version.list;
    instance.listWithHttpInfo = instance._version.listWithHttpInfo;
    instance.getPageWithHttpInfo = function getPageWithHttpInfo(targetUrl, callback) {
        // Use request() directly as it already returns { statusCode, body, headers }
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new TraitGroupPage(instance._version, response, instance._uri, {}, instance._solution),
        }));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
class TraitGroupPage extends TokenPage_1.default {
    /**
     * Initialize the TraitGroupPage
     *
     * @param version - Version of the resource
     * @param response - Response from the API
     * @param uri - URI of the resource
     * @param params - Query parameters
     * @param solution - Path solution
     */
    constructor(version, response, uri, params, solution) {
        super(version, response, uri, params, solution);
    }
    /**
     * Build an instance of TraitGroupInstance
     *
     * @param payload - Payload response from the API
     */
    getInstance(payload) {
        return new TraitGroupInstance(this._version, payload, this._solution.storeId);
    }
    [util_1.inspect.custom](depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.TraitGroupPage = TraitGroupPage;
