// twilio/lib/rest/intelligence/v3/conversation.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Conversational Intelligence API
 * The Conversational Intelligence API includes resources to create and manage intelligence configurations, define and run language operators, and retrieve processed conversations and operator results.  * Use the Configurations resource to create and manage intelligence configurations and define rules that control how and when language operators analyze and transform conversations.  * Use the Operators resource to create custom language operators or retrieve Twilio-author and custom operators. * Use the Conversations resource to retrieve conversations processed with an intelligence configuration, and the OperatorResults resource to retrieve language operator results for those conversations.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConversationPage = exports.ConversationInstance = exports.ConversationContextImpl = exports.Participant = exports.Content = exports.Communication = void 0;
exports.ConversationListInstance = ConversationListInstance;
const util_1 = require("util");
const TokenPage_1 = __importDefault(require("../../../base/TokenPage"));
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
class Communication {
    constructor(payload) {
        this.id = payload["id"];
        this.participantId = payload["participantId"];
        this.content = payload["content"];
        this.createdAt = payload["createdAt"];
    }
}
exports.Communication = Communication;
/**
 * The content of the communication.
 */
class Content {
    constructor(payload) {
        this.text = payload["text"];
    }
}
exports.Content = Content;
class Participant {
    constructor(payload) {
        this.id = payload["id"];
        this.name = payload["name"];
        this.type = payload["type"];
        this.addressValues = payload["addressValues"];
    }
}
exports.Participant = Participant;
class ConversationContextImpl {
    constructor(_version, id) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(id)) {
            throw new Error("Parameter 'id' is not valid.");
        }
        this._solution = { id };
        this._uri = `/Conversations/${id}`;
    }
    fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ConversationInstance(operationVersion, payload, instance._solution.id));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new ConversationInstance(operationVersion, response.body, instance._solution.id),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ConversationContextImpl = ConversationContextImpl;
class ConversationInstance {
    constructor(_version, _payload, id) {
        this._version = _version;
        const payload = _payload;
        this.id = payload.id;
        this.accountId = payload.accountId;
        this.name = payload.name;
        this.status = payload.status;
        this.createdAt = deserialize.iso8601DateTime(payload.createdAt);
        this.updatedAt = deserialize.iso8601DateTime(payload.updatedAt);
        this.intelligenceConfigurationIds = payload.intelligenceConfigurationIds;
        this.conversationConfigurationId = payload.conversationConfigurationId;
        this.channels = payload.channels;
        this.channelIds = payload.channelIds;
        this.participants =
            payload.participants !== null && payload.participants !== undefined
                ? payload.participants.map((payload) => new Participant(payload))
                : null;
        this.communications =
            payload.communications !== null && payload.communications !== undefined
                ? payload.communications.map((payload) => new Communication(payload))
                : null;
        this.operatorResultIds = payload.operatorResultIds;
        this._solution = { id: id };
    }
    get _proxy() {
        this._context =
            this._context ||
                new ConversationContextImpl(this._version, this._solution.id);
        return this._context;
    }
    /**
     * Fetch a ConversationInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ConversationInstance
     */
    fetch(callback) {
        return this._proxy.fetch(callback);
    }
    /**
     * Fetch a ConversationInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed ConversationInstance with HTTP metadata
     */
    fetchWithHttpInfo(callback) {
        return this._proxy.fetchWithHttpInfo(callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            id: this.id,
            accountId: this.accountId,
            name: this.name,
            status: this.status,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
            intelligenceConfigurationIds: this.intelligenceConfigurationIds,
            conversationConfigurationId: this.conversationConfigurationId,
            channels: this.channels,
            channelIds: this.channelIds,
            participants: this.participants,
            communications: this.communications,
            operatorResultIds: this.operatorResultIds,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ConversationInstance = ConversationInstance;
function ConversationListInstance(version) {
    const instance = ((id) => instance.get(id));
    instance.get = function get(id) {
        return new ConversationContextImpl(version, id);
    };
    instance._version = version;
    instance._solution = {};
    instance._uri = `/Conversations`;
    instance.page = function page(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["createdAtBefore"] !== undefined)
            data["createdAtBefore"] = serialize.iso8601DateTime(params["createdAtBefore"]);
        if (params["createdAtAfter"] !== undefined)
            data["createdAtAfter"] = serialize.iso8601DateTime(params["createdAtAfter"]);
        if (params["status"] !== undefined)
            data["status"] = params["status"];
        if (params["channelId"] !== undefined)
            data["channelId"] = params["channelId"];
        if (params["channels"] !== undefined)
            data["channels"] = serialize.map(params["channels"], (e) => e);
        if (params["conversationConfigurationId"] !== undefined)
            data["conversationConfigurationId"] =
                params["conversationConfigurationId"];
        if (params["intelligenceConfigurationIds"] !== undefined)
            data["intelligenceConfigurationIds"] = serialize.map(params["intelligenceConfigurationIds"], (e) => e);
        if (params["operatorIds"] !== undefined)
            data["operatorIds"] = serialize.map(params["operatorIds"], (e) => e);
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.page({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new ConversationPage(operationVersion, payload, instance._uri, data, instance._solution));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.list = instance._version.list;
    instance.getPage = function getPage(targetUrl, callback) {
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((payload) => new ConversationPage(instance._version, payload, instance._uri, {}, instance._solution));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.pageWithHttpInfo = function pageWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["pageSize"] !== undefined)
            data["pageSize"] = params["pageSize"];
        if (params["pageToken"] !== undefined)
            data["pageToken"] = params["pageToken"];
        if (params["createdAtBefore"] !== undefined)
            data["createdAtBefore"] = serialize.iso8601DateTime(params["createdAtBefore"]);
        if (params["createdAtAfter"] !== undefined)
            data["createdAtAfter"] = serialize.iso8601DateTime(params["createdAtAfter"]);
        if (params["status"] !== undefined)
            data["status"] = params["status"];
        if (params["channelId"] !== undefined)
            data["channelId"] = params["channelId"];
        if (params["channels"] !== undefined)
            data["channels"] = serialize.map(params["channels"], (e) => e);
        if (params["conversationConfigurationId"] !== undefined)
            data["conversationConfigurationId"] =
                params["conversationConfigurationId"];
        if (params["intelligenceConfigurationIds"] !== undefined)
            data["intelligenceConfigurationIds"] = serialize.map(params["intelligenceConfigurationIds"], (e) => e);
        if (params["operatorIds"] !== undefined)
            data["operatorIds"] = serialize.map(params["operatorIds"], (e) => e);
        const headers = {};
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // For page operations, use page() directly as it already returns { statusCode, body, headers }
        // IMPORTANT: Pass full response to Page constructor, not response.body
        let operationPromise = operationVersion
            .page({ uri: instance._uri, method: "get", params: data, headers })
            .then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new ConversationPage(operationVersion, response, instance._uri, data, instance._solution),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.each = instance._version.each;
    instance.eachWithHttpInfo = instance._version.eachWithHttpInfo;
    instance.list = instance._version.list;
    instance.listWithHttpInfo = instance._version.listWithHttpInfo;
    instance.getPageWithHttpInfo = function getPageWithHttpInfo(targetUrl, callback) {
        // Use request() directly as it already returns { statusCode, body, headers }
        const operationPromise = instance._version._domain.twilio.request({
            method: "get",
            uri: targetUrl,
        });
        let pagePromise = operationPromise.then((response) => ({
            statusCode: response.statusCode,
            headers: response.headers,
            body: new ConversationPage(instance._version, response, instance._uri, {}, instance._solution),
        }));
        pagePromise = instance._version.setPromiseCallback(pagePromise, callback);
        return pagePromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
class ConversationPage extends TokenPage_1.default {
    /**
     * Initialize the ConversationPage
     *
     * @param version - Version of the resource
     * @param response - Response from the API
     * @param uri - URI of the resource
     * @param params - Query parameters
     * @param solution - Path solution
     */
    constructor(version, response, uri, params, solution) {
        super(version, response, uri, params, solution);
    }
    /**
     * Build an instance of ConversationInstance
     *
     * @param payload - Payload response from the API
     */
    getInstance(payload) {
        return new ConversationInstance(this._version, payload);
    }
    [util_1.inspect.custom](depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.ConversationPage = ConversationPage;
