// twilio/lib/rest/intelligence/V3.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Conversational Intelligence API
 * The Conversational Intelligence API includes resources to create and manage intelligence configurations, define and run language operators, and retrieve processed conversations and operator results.  * Use the Configurations resource to create and manage intelligence configurations and define rules that control how and when language operators analyze and transform conversations.  * Use the Operators resource to create custom language operators or retrieve Twilio-author and custom operators. * Use the Conversations resource to retrieve conversations processed with an intelligence configuration, and the OperatorResults resource to retrieve language operator results for those conversations.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Version_1 = __importDefault(require("../../base/Version"));
const configuration_1 = require("./v3/configuration");
const conversation_1 = require("./v3/conversation");
const operator_1 = require("./v3/operator");
const operatorResult_1 = require("./v3/operatorResult");
const version_1 = require("./v3/version");
class V3 extends Version_1.default {
    /**
     * Initialize the V3 version of Intelligence
     *
     * @param domain - The Twilio (Twilio.Intelligence) domain
     */
    constructor(domain) {
        super(domain, "v3");
    }
    /** Getter for configurations resource */
    get configurations() {
        this._configurations =
            this._configurations || (0, configuration_1.ConfigurationListInstance)(this);
        return this._configurations;
    }
    /** Getter for conversations resource */
    get conversations() {
        this._conversations = this._conversations || (0, conversation_1.ConversationListInstance)(this);
        return this._conversations;
    }
    /** Getter for operators resource */
    get operators() {
        this._operators = this._operators || (0, operator_1.OperatorListInstance)(this);
        return this._operators;
    }
    /** Getter for operatorResults resource */
    get operatorResults() {
        this._operatorResults =
            this._operatorResults || (0, operatorResult_1.OperatorResultListInstance)(this);
        return this._operatorResults;
    }
    /** Implementation */
    versions(id, version) {
        const listInstance = (0, version_1.VersionListInstance)(this, id);
        if (version !== undefined) {
            return listInstance.get(version);
        }
        return listInstance;
    }
}
exports.default = V3;
