// twilio/lib/rest/numbers/v1/portingPortIn.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio - Numbers
 * This is the public Twilio REST API.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortingPortInInstance = exports.PortingPortInContextImpl = exports.NumbersV1PortingPortInPhoneNumberResult = exports.NumbersV1PortingPortInCreatePhoneNumbers = exports.NumbersV1PortingPortInCreate = exports.NumbersV1PortingLosingCarrierInformation = exports.NumbersV1PortingAddress = void 0;
exports.PortingPortInListInstance = PortingPortInListInstance;
const util_1 = require("util");
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
class NumbersV1PortingAddress {
    constructor(payload) {
        this.street = payload["street"];
        this.street2 = payload["street_2"];
        this.city = payload["city"];
        this.state = payload["state"];
        this.zip = payload["zip"];
        this.country = payload["country"];
    }
}
exports.NumbersV1PortingAddress = NumbersV1PortingAddress;
class NumbersV1PortingLosingCarrierInformation {
    constructor(payload) {
        this.customerName = payload["customer_name"];
        this.accountNumber = payload["account_number"];
        this.accountTelephoneNumber = payload["account_telephone_number"];
        this.addressSid = payload["address_sid"];
        this.address = payload["address"];
        this.authorizedRepresentative = payload["authorized_representative"];
        this.authorizedRepresentativeEmail =
            payload["authorized_representative_email"];
        this.customerType = payload["customer_type"];
        this.authorizedRepresentativeKatakana =
            payload["authorized_representative_katakana"];
        this.subMunicipality = payload["sub_municipality"];
        this.building = payload["building"];
        this.katakanaName = payload["katakana_name"];
    }
}
exports.NumbersV1PortingLosingCarrierInformation = NumbersV1PortingLosingCarrierInformation;
class NumbersV1PortingPortInCreate {
    constructor(payload) {
        this.accountSid = payload["account_sid"];
        this.documents = payload["documents"];
        this.phoneNumbers = payload["phone_numbers"];
        this.losingCarrierInformation = payload["losing_carrier_information"];
        this.notificationEmails = payload["notification_emails"];
        this.targetPortInDate = payload["target_port_in_date"];
        this.targetPortInTimeRangeStart =
            payload["target_port_in_time_range_start"];
        this.targetPortInTimeRangeEnd = payload["target_port_in_time_range_end"];
        this.bundleSid = payload["bundle_sid"];
        this.portabilityAdvanceCarrier = payload["portability_advance_carrier"];
        this.autoCancelApprovalNumbers = payload["auto_cancel_approval_numbers"];
    }
}
exports.NumbersV1PortingPortInCreate = NumbersV1PortingPortInCreate;
class NumbersV1PortingPortInCreatePhoneNumbers {
    constructor(payload) {
        this.phoneNumber = payload["phone_number"];
        this.pin = payload["pin"];
    }
}
exports.NumbersV1PortingPortInCreatePhoneNumbers = NumbersV1PortingPortInCreatePhoneNumbers;
class NumbersV1PortingPortInPhoneNumberResult {
    constructor(payload) {
        this.notPortabilityReason = payload["not_portability_reason"];
        this.notPortabilityReasonCode = payload["not_portability_reason_code"];
        this.numberType = payload["number_type"];
        this.phoneNumber = payload["phone_number"];
        this.portDate = payload["port_date"];
        this.portInPhoneNumberSid = payload["port_in_phone_number_sid"];
        this.portInPhoneNumberStatus = payload["port_in_phone_number_status"];
        this.portable = payload["portable"];
        this.rejectionReason = payload["rejection_reason"];
        this.rejectionReasonCode = payload["rejection_reason_code"];
        this.statusLastTimeUpdatedTimestamp =
            payload["status_last_time_updated_timestamp"];
        this.externalPortingVendorPhoneNumberId =
            payload["external_porting_vendor_phone_number_id"];
    }
}
exports.NumbersV1PortingPortInPhoneNumberResult = NumbersV1PortingPortInPhoneNumberResult;
class PortingPortInContextImpl {
    constructor(_version, portInRequestSid) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(portInRequestSid)) {
            throw new Error("Parameter 'portInRequestSid' is not valid.");
        }
        this._solution = { portInRequestSid };
        this._uri = `/Porting/PortIn/${portInRequestSid}`;
    }
    remove(callback) {
        const headers = {};
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.remove({
            uri: instance._uri,
            method: "delete",
            headers,
        });
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    removeWithHttpInfo(callback) {
        const headers = {};
        const instance = this;
        let operationVersion = instance._version;
        // DELETE operation - returns boolean based on status code
        let operationPromise = operationVersion
            .removeWithResponseInfo({ uri: instance._uri, method: "delete", headers })
            .then((response) => ({
            ...response,
            body: response.statusCode === 204,
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new PortingPortInInstance(operationVersion, payload, instance._solution.portInRequestSid));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new PortingPortInInstance(operationVersion, response.body, instance._solution.portInRequestSid),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.PortingPortInContextImpl = PortingPortInContextImpl;
class PortingPortInInstance {
    constructor(_version, payload, portInRequestSid) {
        this._version = _version;
        this.portInRequestSid = payload.port_in_request_sid;
        this.url = payload.url;
        this.accountSid = payload.account_sid;
        this.notificationEmails = payload.notification_emails;
        this.targetPortInDate = deserialize.iso8601Date(payload.target_port_in_date);
        this.targetPortInTimeRangeStart = payload.target_port_in_time_range_start;
        this.targetPortInTimeRangeEnd = payload.target_port_in_time_range_end;
        this.portInRequestStatus = payload.port_in_request_status;
        this.orderCancellationReason = payload.order_cancellation_reason;
        this.losingCarrierInformation =
            payload.losing_carrier_information !== null &&
                payload.losing_carrier_information !== undefined
                ? new NumbersV1PortingLosingCarrierInformation(payload.losing_carrier_information)
                : null;
        this.phoneNumbers =
            payload.phone_numbers !== null && payload.phone_numbers !== undefined
                ? payload.phone_numbers.map((payload) => new NumbersV1PortingPortInPhoneNumberResult(payload))
                : null;
        this.bundleSid = payload.bundle_sid;
        this.portabilityAdvanceCarrier = payload.portability_advance_carrier;
        this.autoCancelApprovalNumbers = payload.auto_cancel_approval_numbers;
        this.documents = payload.documents;
        this.dateCreated = deserialize.iso8601DateTime(payload.date_created);
        this.supportTicketId = deserialize.integer(payload.support_ticket_id);
        this.signatureRequestUrl = payload.signature_request_url;
        this._solution = {
            portInRequestSid: portInRequestSid || this.portInRequestSid,
        };
    }
    get _proxy() {
        this._context =
            this._context ||
                new PortingPortInContextImpl(this._version, this._solution.portInRequestSid);
        return this._context;
    }
    /**
     * Remove a PortingPortInInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed boolean
     */
    remove(callback) {
        return this._proxy.remove(callback);
    }
    /**
     * Remove a PortingPortInInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed boolean with HTTP metadata
     */
    removeWithHttpInfo(callback) {
        return this._proxy.removeWithHttpInfo(callback);
    }
    /**
     * Fetch a PortingPortInInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed PortingPortInInstance
     */
    fetch(callback) {
        return this._proxy.fetch(callback);
    }
    /**
     * Fetch a PortingPortInInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed PortingPortInInstance with HTTP metadata
     */
    fetchWithHttpInfo(callback) {
        return this._proxy.fetchWithHttpInfo(callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            portInRequestSid: this.portInRequestSid,
            url: this.url,
            accountSid: this.accountSid,
            notificationEmails: this.notificationEmails,
            targetPortInDate: this.targetPortInDate,
            targetPortInTimeRangeStart: this.targetPortInTimeRangeStart,
            targetPortInTimeRangeEnd: this.targetPortInTimeRangeEnd,
            portInRequestStatus: this.portInRequestStatus,
            orderCancellationReason: this.orderCancellationReason,
            losingCarrierInformation: this.losingCarrierInformation,
            phoneNumbers: this.phoneNumbers,
            bundleSid: this.bundleSid,
            portabilityAdvanceCarrier: this.portabilityAdvanceCarrier,
            autoCancelApprovalNumbers: this.autoCancelApprovalNumbers,
            documents: this.documents,
            dateCreated: this.dateCreated,
            supportTicketId: this.supportTicketId,
            signatureRequestUrl: this.signatureRequestUrl,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.PortingPortInInstance = PortingPortInInstance;
function PortingPortInListInstance(version) {
    const instance = ((portInRequestSid) => instance.get(portInRequestSid));
    instance.get = function get(portInRequestSid) {
        return new PortingPortInContextImpl(version, portInRequestSid);
    };
    instance._version = version;
    instance._solution = {};
    instance._uri = `/Porting/PortIn`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new PortingPortInInstance(operationVersion, payload));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new PortingPortInInstance(operationVersion, response.body),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
