// twilio/lib/rest/numbers/v1/senderIdRegistration.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio - Numbers
 * This is the public Twilio REST API.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SenderIdRegistrationInstance = exports.NumbersV1EmbeddedSession = exports.NumbersV1CreateEmbeddedRegistrationRequest = void 0;
exports.SenderIdRegistrationListInstance = SenderIdRegistrationListInstance;
const util_1 = require("util");
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
class NumbersV1CreateEmbeddedRegistrationRequest {
    constructor(payload) {
        this.regulationId = payload["regulationId"];
        this.regulationVersion = payload["regulationVersion"];
        this.friendlyName = payload["friendlyName"];
        this.statusNotificationEmail = payload["statusNotificationEmail"];
        this.statusCallbackUrl = payload["statusCallbackUrl"];
        this.comments = payload["comments"];
        this.themeSetId = payload["themeSetId"];
        this.data = payload["data"];
    }
}
exports.NumbersV1CreateEmbeddedRegistrationRequest = NumbersV1CreateEmbeddedRegistrationRequest;
class NumbersV1EmbeddedSession {
    constructor(payload) {
        this.sessionId = payload["sessionId"];
        this.sessionToken = payload["sessionToken"];
    }
}
exports.NumbersV1EmbeddedSession = NumbersV1EmbeddedSession;
function SenderIdRegistrationListInstance(version) {
    const instance = {};
    instance._version = version;
    instance._solution = {};
    instance._uri = `/SenderIdRegistrations`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new SenderIdRegistrationInstance(operationVersion, payload));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new SenderIdRegistrationInstance(operationVersion, response.body),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
class SenderIdRegistrationInstance {
    constructor(_version, payload) {
        this._version = _version;
        this.id = payload.id;
        this.regulationId = payload.regulationId;
        this.regulationVersion = deserialize.integer(payload.regulationVersion);
        this.friendlyName = payload.friendlyName;
        this.status = payload.status;
        this.statusNotificationEmail = payload.statusNotificationEmail;
        this.statusCallbackUrl = payload.statusCallbackUrl;
        this.comments = payload.comments;
        this.embeddedSession =
            payload.embeddedSession !== null && payload.embeddedSession !== undefined
                ? new NumbersV1EmbeddedSession(payload.embeddedSession)
                : null;
        this.data = payload.data;
        this.dateCreated = deserialize.iso8601DateTime(payload.dateCreated);
        this.dateUpdated = deserialize.iso8601DateTime(payload.dateUpdated);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            id: this.id,
            regulationId: this.regulationId,
            regulationVersion: this.regulationVersion,
            friendlyName: this.friendlyName,
            status: this.status,
            statusNotificationEmail: this.statusNotificationEmail,
            statusCallbackUrl: this.statusCallbackUrl,
            comments: this.comments,
            embeddedSession: this.embeddedSession,
            data: this.data,
            dateCreated: this.dateCreated,
            dateUpdated: this.dateUpdated,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.SenderIdRegistrationInstance = SenderIdRegistrationInstance;
