// twilio/lib/rest/lookups/v2/phoneNumber.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio - Lookups
 * This is the public Twilio REST API.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhoneNumberInstance = exports.PhoneNumberContextImpl = exports.SmsPumpingRiskInfo = exports.SimSwapInfo = exports.ReassignedNumberInfo = exports.LineTypeIntelligenceInfo = exports.LineStatusInfo = exports.LastSimSwapInfo = exports.IdentityMatchInfo = exports.CallerNameInfo = exports.CallForwardingInfo = void 0;
exports.PhoneNumberListInstance = PhoneNumberListInstance;
const util_1 = require("util");
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
class CallForwardingInfo {
    constructor(payload) {
        this.callForwardingEnabled = payload["call_forwarding_enabled"];
        this.errorCode = payload["error_code"];
    }
}
exports.CallForwardingInfo = CallForwardingInfo;
class CallerNameInfo {
    constructor(payload) {
        this.callerName = payload["caller_name"];
        this.callerType = payload["caller_type"];
        this.errorCode = payload["error_code"];
    }
}
exports.CallerNameInfo = CallerNameInfo;
class IdentityMatchInfo {
    constructor(payload) {
        this.firstNameMatch = payload["first_name_match"];
        this.lastNameMatch = payload["last_name_match"];
        this.addressLinesMatch = payload["address_lines_match"];
        this.cityMatch = payload["city_match"];
        this.stateMatch = payload["state_match"];
        this.postalCodeMatch = payload["postal_code_match"];
        this.addressCountryMatch = payload["address_country_match"];
        this.nationalIdMatch = payload["national_id_match"];
        this.dateOfBirthMatch = payload["date_of_birth_match"];
        this.summaryScore = payload["summary_score"];
        this.errorCode = payload["error_code"];
        this.errorMessage = payload["error_message"];
    }
}
exports.IdentityMatchInfo = IdentityMatchInfo;
class LastSimSwapInfo {
    constructor(payload) {
        this.lastSimSwapDate = payload["last_sim_swap_date"];
        this.swappedPeriod = payload["swapped_period"];
        this.swappedInPeriod = payload["swapped_in_period"];
    }
}
exports.LastSimSwapInfo = LastSimSwapInfo;
class LineStatusInfo {
    constructor(payload) {
        this.status = payload["status"];
        this.errorCode = payload["error_code"];
    }
}
exports.LineStatusInfo = LineStatusInfo;
class LineTypeIntelligenceInfo {
    constructor(payload) {
        this.mobileCountryCode = payload["mobile_country_code"];
        this.mobileNetworkCode = payload["mobile_network_code"];
        this.carrierName = payload["carrier_name"];
        this.type = payload["type"];
        this.errorCode = payload["error_code"];
    }
}
exports.LineTypeIntelligenceInfo = LineTypeIntelligenceInfo;
class ReassignedNumberInfo {
    constructor(payload) {
        this.lastVerifiedDate = payload["last_verified_date"];
        this.isNumberReassigned = payload["is_number_reassigned"];
        this.errorCode = payload["error_code"];
    }
}
exports.ReassignedNumberInfo = ReassignedNumberInfo;
class SimSwapInfo {
    constructor(payload) {
        this.lastSimSwap = payload["last_sim_swap"];
        this.carrierName = payload["carrier_name"];
        this.mobileCountryCode = payload["mobile_country_code"];
        this.mobileNetworkCode = payload["mobile_network_code"];
        this.errorCode = payload["error_code"];
    }
}
exports.SimSwapInfo = SimSwapInfo;
class SmsPumpingRiskInfo {
    constructor(payload) {
        this.carrierRiskCategory = payload["carrier_risk_category"];
        this.numberBlocked = payload["number_blocked"];
        this.numberBlockedDate = payload["number_blocked_date"];
        this.numberBlockedLast3Months = payload["number_blocked_last_3_months"];
        this.smsPumpingRiskScore = payload["sms_pumping_risk_score"];
        this.errorCode = payload["error_code"];
    }
}
exports.SmsPumpingRiskInfo = SmsPumpingRiskInfo;
class PhoneNumberContextImpl {
    constructor(_version, phoneNumber) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(phoneNumber)) {
            throw new Error("Parameter 'phoneNumber' is not valid.");
        }
        this._solution = { phoneNumber };
        this._uri = `/PhoneNumbers/${phoneNumber}`;
    }
    fetch(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["fields"] !== undefined)
            data["Fields"] = params["fields"];
        if (params["countryCode"] !== undefined)
            data["CountryCode"] = params["countryCode"];
        if (params["firstName"] !== undefined)
            data["FirstName"] = params["firstName"];
        if (params["lastName"] !== undefined)
            data["LastName"] = params["lastName"];
        if (params["addressLine1"] !== undefined)
            data["AddressLine1"] = params["addressLine1"];
        if (params["addressLine2"] !== undefined)
            data["AddressLine2"] = params["addressLine2"];
        if (params["city"] !== undefined)
            data["City"] = params["city"];
        if (params["state"] !== undefined)
            data["State"] = params["state"];
        if (params["postalCode"] !== undefined)
            data["PostalCode"] = params["postalCode"];
        if (params["addressCountryCode"] !== undefined)
            data["AddressCountryCode"] = params["addressCountryCode"];
        if (params["nationalId"] !== undefined)
            data["NationalId"] = params["nationalId"];
        if (params["dateOfBirth"] !== undefined)
            data["DateOfBirth"] = params["dateOfBirth"];
        if (params["lastVerifiedDate"] !== undefined)
            data["LastVerifiedDate"] = params["lastVerifiedDate"];
        if (params["verificationSid"] !== undefined)
            data["VerificationSid"] = params["verificationSid"];
        if (params["partnerSubId"] !== undefined)
            data["PartnerSubId"] = params["partnerSubId"];
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new PhoneNumberInstance(operationVersion, payload, instance._solution.phoneNumber));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(params, callback) {
        if (params instanceof Function) {
            callback = params;
            params = {};
        }
        else {
            params = params || {};
        }
        let data = {};
        if (params["fields"] !== undefined)
            data["Fields"] = params["fields"];
        if (params["countryCode"] !== undefined)
            data["CountryCode"] = params["countryCode"];
        if (params["firstName"] !== undefined)
            data["FirstName"] = params["firstName"];
        if (params["lastName"] !== undefined)
            data["LastName"] = params["lastName"];
        if (params["addressLine1"] !== undefined)
            data["AddressLine1"] = params["addressLine1"];
        if (params["addressLine2"] !== undefined)
            data["AddressLine2"] = params["addressLine2"];
        if (params["city"] !== undefined)
            data["City"] = params["city"];
        if (params["state"] !== undefined)
            data["State"] = params["state"];
        if (params["postalCode"] !== undefined)
            data["PostalCode"] = params["postalCode"];
        if (params["addressCountryCode"] !== undefined)
            data["AddressCountryCode"] = params["addressCountryCode"];
        if (params["nationalId"] !== undefined)
            data["NationalId"] = params["nationalId"];
        if (params["dateOfBirth"] !== undefined)
            data["DateOfBirth"] = params["dateOfBirth"];
        if (params["lastVerifiedDate"] !== undefined)
            data["LastVerifiedDate"] = params["lastVerifiedDate"];
        if (params["verificationSid"] !== undefined)
            data["VerificationSid"] = params["verificationSid"];
        if (params["partnerSubId"] !== undefined)
            data["PartnerSubId"] = params["partnerSubId"];
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            params: data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new PhoneNumberInstance(operationVersion, response.body, instance._solution.phoneNumber),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.PhoneNumberContextImpl = PhoneNumberContextImpl;
class PhoneNumberInstance {
    constructor(_version, payload, phoneNumber) {
        this._version = _version;
        this.callingCountryCode = payload.calling_country_code;
        this.countryCode = payload.country_code;
        this.phoneNumber = payload.phone_number;
        this.nationalFormat = payload.national_format;
        this.valid = payload.valid;
        this.validationErrors = payload.validation_errors;
        this.callerName =
            payload.caller_name !== null && payload.caller_name !== undefined
                ? new CallerNameInfo(payload.caller_name)
                : null;
        this.simSwap =
            payload.sim_swap !== null && payload.sim_swap !== undefined
                ? new SimSwapInfo(payload.sim_swap)
                : null;
        this.callForwarding =
            payload.call_forwarding !== null && payload.call_forwarding !== undefined
                ? new CallForwardingInfo(payload.call_forwarding)
                : null;
        this.lineTypeIntelligence =
            payload.line_type_intelligence !== null &&
                payload.line_type_intelligence !== undefined
                ? new LineTypeIntelligenceInfo(payload.line_type_intelligence)
                : null;
        this.lineStatus =
            payload.line_status !== null && payload.line_status !== undefined
                ? new LineStatusInfo(payload.line_status)
                : null;
        this.identityMatch =
            payload.identity_match !== null && payload.identity_match !== undefined
                ? new IdentityMatchInfo(payload.identity_match)
                : null;
        this.reassignedNumber =
            payload.reassigned_number !== null &&
                payload.reassigned_number !== undefined
                ? new ReassignedNumberInfo(payload.reassigned_number)
                : null;
        this.smsPumpingRisk =
            payload.sms_pumping_risk !== null &&
                payload.sms_pumping_risk !== undefined
                ? new SmsPumpingRiskInfo(payload.sms_pumping_risk)
                : null;
        this.phoneNumberQualityScore = payload.phone_number_quality_score;
        this.preFill = payload.pre_fill;
        this.url = payload.url;
        this._solution = { phoneNumber: phoneNumber || this.phoneNumber };
    }
    get _proxy() {
        this._context =
            this._context ||
                new PhoneNumberContextImpl(this._version, this._solution.phoneNumber);
        return this._context;
    }
    fetch(params, callback) {
        return this._proxy.fetch(params, callback);
    }
    fetchWithHttpInfo(params, callback) {
        return this._proxy.fetchWithHttpInfo(params, callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            callingCountryCode: this.callingCountryCode,
            countryCode: this.countryCode,
            phoneNumber: this.phoneNumber,
            nationalFormat: this.nationalFormat,
            valid: this.valid,
            validationErrors: this.validationErrors,
            callerName: this.callerName,
            simSwap: this.simSwap,
            callForwarding: this.callForwarding,
            lineTypeIntelligence: this.lineTypeIntelligence,
            lineStatus: this.lineStatus,
            identityMatch: this.identityMatch,
            reassignedNumber: this.reassignedNumber,
            smsPumpingRisk: this.smsPumpingRisk,
            phoneNumberQualityScore: this.phoneNumberQualityScore,
            preFill: this.preFill,
            url: this.url,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.PhoneNumberInstance = PhoneNumberInstance;
function PhoneNumberListInstance(version) {
    const instance = ((phoneNumber) => instance.get(phoneNumber));
    instance.get = function get(phoneNumber) {
        return new PhoneNumberContextImpl(version, phoneNumber);
    };
    instance._version = version;
    instance._solution = {};
    instance._uri = ``;
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
