// twilio/lib/rest/memory/V1.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio Memory API
 * APIs for managing memory stores, profiles, events, and conversational intelligence capabilities.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Version_1 = __importDefault(require("../../base/Version"));
const bulk_1 = require("./v1/bulk");
const conversationSummary_1 = require("./v1/conversationSummary");
const dataMapping_1 = require("./v1/dataMapping");
const event_1 = require("./v1/event");
const identifier_1 = require("./v1/identifier");
const identityResolutionSetting_1 = require("./v1/identityResolutionSetting");
const import_1 = require("./v1/import");
const lookup_1 = require("./v1/lookup");
const observation_1 = require("./v1/observation");
const operation_1 = require("./v1/operation");
const profile_1 = require("./v1/profile");
const recall_1 = require("./v1/recall");
const revision_1 = require("./v1/revision");
const store_1 = require("./v1/store");
const trait_1 = require("./v1/trait");
const traitGroup_1 = require("./v1/traitGroup");
class V1 extends Version_1.default {
    /**
     * Initialize the V1 version of Memory
     *
     * @param domain - The Twilio (Twilio.Memory) domain
     */
    constructor(domain) {
        super(domain, "v1");
    }
    /** Accessor for bulk resource */
    bulk(storeId) {
        return (0, bulk_1.BulkListInstance)(this)(storeId);
    }
    /** Implementation */
    conversationSummaries(storeId, profileId, summaryId) {
        const listInstance = (0, conversationSummary_1.ConversationSummaryListInstance)(this, storeId, profileId);
        if (summaryId !== undefined) {
            return listInstance.get(summaryId);
        }
        return listInstance;
    }
    /** Implementation */
    dataMappings(storeId, dataMappingId) {
        const listInstance = (0, dataMapping_1.DataMappingListInstance)(this, storeId);
        if (dataMappingId !== undefined) {
            return listInstance.get(dataMappingId);
        }
        return listInstance;
    }
    /** Accessor for events resource */
    events(storeId, profileId) {
        return (0, event_1.EventListInstance)(this, storeId, profileId);
    }
    /** Implementation */
    identifiers(storeId, profileId, idType) {
        const listInstance = (0, identifier_1.IdentifierListInstance)(this, storeId, profileId);
        if (idType !== undefined) {
            return listInstance.get(idType);
        }
        return listInstance;
    }
    /** Accessor for identityResolutionSettings resource */
    identityResolutionSettings(storeId) {
        return (0, identityResolutionSetting_1.IdentityResolutionSettingListInstance)(this)(storeId);
    }
    /** Implementation */
    imports(storeId, importId) {
        const listInstance = (0, import_1.ImportListInstance)(this, storeId);
        if (importId !== undefined) {
            return listInstance.get(importId);
        }
        return listInstance;
    }
    /** Accessor for lookup resource */
    lookup(storeId) {
        return (0, lookup_1.LookupListInstance)(this, storeId);
    }
    /** Implementation */
    observations(storeId, profileId, observationId) {
        const listInstance = (0, observation_1.ObservationListInstance)(this, storeId, profileId);
        if (observationId !== undefined) {
            return listInstance.get(observationId);
        }
        return listInstance;
    }
    /** Getter for operations resource */
    get operations() {
        this._operations = this._operations || (0, operation_1.OperationListInstance)(this);
        return this._operations;
    }
    /** Implementation */
    profiles(storeId, profileId) {
        const listInstance = (0, profile_1.ProfileListInstance)(this, storeId);
        if (profileId !== undefined) {
            return listInstance.get(profileId);
        }
        return listInstance;
    }
    /** Accessor for recall resource */
    recall(storeId, profileId) {
        return (0, recall_1.RecallListInstance)(this, storeId, profileId);
    }
    /** Accessor for revisions resource */
    revisions(storeId, profileId, observationId) {
        return (0, revision_1.RevisionListInstance)(this, storeId, profileId, observationId);
    }
    /** Getter for stores resource */
    get stores() {
        this._stores = this._stores || (0, store_1.StoreListInstance)(this);
        return this._stores;
    }
    /** Accessor for traits resource */
    traits(storeId, profileId) {
        return (0, trait_1.TraitListInstance)(this, storeId, profileId);
    }
    /** Implementation */
    traitGroups(storeId, traitGroupName) {
        const listInstance = (0, traitGroup_1.TraitGroupListInstance)(this, storeId);
        if (traitGroupName !== undefined) {
            return listInstance.get(traitGroupName);
        }
        return listInstance;
    }
}
exports.default = V1;
