// twilio/lib/rest/insights/V2.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Sample/reference Twilio API.
 * This is the reference API for the rest-proxy server.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Version_1 = __importDefault(require("../../base/Version"));
const inbound_1 = require("./v2/inbound");
const outbound_1 = require("./v2/outbound");
const report_1 = require("./v2/report");
class V2 extends Version_1.default {
    /**
     * Initialize the V2 version of Insights
     *
     * @param domain - The Twilio (Twilio.Insights) domain
     */
    constructor(domain) {
        super(domain, "v2");
    }
    /** Accessor for inbound resource */
    inbound(reportId) {
        return (0, inbound_1.InboundListInstance)(this, reportId);
    }
    /** Accessor for outbound resource */
    outbound(reportId) {
        return (0, outbound_1.OutboundListInstance)(this, reportId);
    }
    /** Getter for reports resource */
    get reports() {
        this._reports = this._reports || (0, report_1.ReportListInstance)(this);
        return this._reports;
    }
}
exports.default = V2;
