// twilio/lib/rest/voice/v3/transcription.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio API definition for public-api voice
 * Powers Twilio public-api voice
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranscriptionInstance = exports.TranscriptionContextImpl = exports.VoiceV3TranscriptionTranscriptionStatusCallback = exports.VoiceV3TranscriptionTranscription = exports.VoiceV3TranscriptionResolvedConfigurationParticipantDefaults = exports.VoiceV3TranscriptionResolvedConfiguration = exports.VoiceV3TranscriptionParticipant = exports.CreateV3TranscriptionsRequest = void 0;
exports.TranscriptionListInstance = TranscriptionListInstance;
const util_1 = require("util");
const deserialize = require("../../../base/deserialize");
const serialize = require("../../../base/serialize");
const utility_1 = require("../../../base/utility");
class CreateV3TranscriptionsRequest {
    constructor(payload) {
        this.transcriptionConfigurationId = payload["transcriptionConfigurationId"];
        this.inputSource = payload["inputSource"];
        this.sourceId = payload["sourceId"];
        this.participants = payload["participants"];
        this.mediaUrl = payload["mediaUrl"];
        this.audioStartedAt = payload["audioStartedAt"];
    }
}
exports.CreateV3TranscriptionsRequest = CreateV3TranscriptionsRequest;
class VoiceV3TranscriptionParticipant {
    constructor(payload) {
        this.type = payload["type"];
        this.address = payload["address"];
        this.name = payload["name"];
        this.audioChannelIndex = payload["audioChannelIndex"];
    }
}
exports.VoiceV3TranscriptionParticipant = VoiceV3TranscriptionParticipant;
class VoiceV3TranscriptionResolvedConfiguration {
    constructor(payload) {
        this.transcriptionEngine = payload["transcriptionEngine"];
        this.speechModel = payload["speechModel"];
        this.language = payload["language"];
        this.transcriptionStatusCallback = payload["transcriptionStatusCallback"];
        this.conversationConfigurationId = payload["conversationConfigurationId"];
        this.participantDefaults = payload["participantDefaults"];
    }
}
exports.VoiceV3TranscriptionResolvedConfiguration = VoiceV3TranscriptionResolvedConfiguration;
class VoiceV3TranscriptionResolvedConfigurationParticipantDefaults {
    constructor(payload) {
        this.audioChannelIndex = payload["audioChannelIndex"];
        this.type = payload["type"];
    }
}
exports.VoiceV3TranscriptionResolvedConfigurationParticipantDefaults = VoiceV3TranscriptionResolvedConfigurationParticipantDefaults;
class VoiceV3TranscriptionTranscription {
    constructor(payload) {
        this.id = payload["id"];
        this.accountId = payload["accountId"];
        this.status = payload["status"];
        this.transcriptionConfigurationId = payload["transcriptionConfigurationId"];
        this.mediaUrl = payload["mediaUrl"];
        this.sourceId = payload["sourceId"];
        this.audioStartedAt = payload["audioStartedAt"];
        this.conversationId = payload["conversationId"];
        this.participants = payload["participants"];
        this.duration = payload["duration"];
        this.resolvedConfiguration = payload["resolvedConfiguration"];
        this.createdAt = payload["createdAt"];
        this.updatedAt = payload["updatedAt"];
        this.url = payload["url"];
    }
}
exports.VoiceV3TranscriptionTranscription = VoiceV3TranscriptionTranscription;
class VoiceV3TranscriptionTranscriptionStatusCallback {
    constructor(payload) {
        this.url = payload["url"];
        this.method = payload["method"];
        this.events = payload["events"];
    }
}
exports.VoiceV3TranscriptionTranscriptionStatusCallback = VoiceV3TranscriptionTranscriptionStatusCallback;
class TranscriptionContextImpl {
    constructor(_version, transcriptionId) {
        this._version = _version;
        if (!(0, utility_1.isValidPathParam)(transcriptionId)) {
            throw new Error("Parameter 'transcriptionId' is not valid.");
        }
        this._solution = { transcriptionId };
        this._uri = `/Transcriptions/${transcriptionId}`;
    }
    fetch(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version, operationPromise = operationVersion.fetch({
            uri: instance._uri,
            method: "get",
            headers,
        });
        operationPromise = operationPromise.then((payload) => new TranscriptionInstance(operationVersion, payload, instance._solution.transcriptionId));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    fetchWithHttpInfo(callback) {
        const headers = {};
        headers["Accept"] = "application/json";
        const instance = this;
        let operationVersion = instance._version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .fetchWithResponseInfo({
            uri: instance._uri,
            method: "get",
            headers,
        })
            .then((response) => ({
            ...response,
            body: new TranscriptionInstance(operationVersion, response.body, instance._solution.transcriptionId),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return this._solution;
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.TranscriptionContextImpl = TranscriptionContextImpl;
/**
 * Response envelope for long-running operations (202 Accepted pattern). Returned immediately on acceptance and on each status poll. Extensible to allow additional fields in future versions.
 */
class TranscriptionInstance {
    constructor(_version, _payload, transcriptionId) {
        this._version = _version;
        const payload = _payload;
        this.status = payload.status;
        this.statusUrl = payload.statusUrl;
        this.transcription =
            payload.transcription !== null && payload.transcription !== undefined
                ? new VoiceV3TranscriptionTranscription(payload.transcription)
                : null;
        this.operationId = payload.operationId;
        this._solution = { transcriptionId: transcriptionId };
    }
    get _proxy() {
        this._context =
            this._context ||
                new TranscriptionContextImpl(this._version, this._solution.transcriptionId);
        return this._context;
    }
    /**
     * Fetch a TranscriptionInstance
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed TranscriptionInstance
     */
    fetch(callback) {
        return this._proxy.fetch(callback);
    }
    /**
     * Fetch a TranscriptionInstance and return HTTP info
     *
     * @param callback - Callback to handle processed record
     *
     * @returns Resolves to processed TranscriptionInstance with HTTP metadata
     */
    fetchWithHttpInfo(callback) {
        return this._proxy.fetchWithHttpInfo(callback);
    }
    /**
     * Provide a user-friendly representation
     *
     * @returns Object
     */
    toJSON() {
        return {
            status: this.status,
            statusUrl: this.statusUrl,
            transcription: this.transcription,
            operationId: this.operationId,
        };
    }
    [util_1.inspect.custom](_depth, options) {
        return (0, util_1.inspect)(this.toJSON(), options);
    }
}
exports.TranscriptionInstance = TranscriptionInstance;
function TranscriptionListInstance(version) {
    const instance = ((transcriptionId) => instance.get(transcriptionId));
    instance.get = function get(transcriptionId) {
        return new TranscriptionContextImpl(version, transcriptionId);
    };
    instance._version = version;
    instance._solution = {};
    instance._uri = `/Transcriptions`;
    instance.create = function create(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version, operationPromise = operationVersion.create({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        });
        operationPromise = operationPromise.then((payload) => new TranscriptionInstance(operationVersion, payload));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.createWithHttpInfo = function createWithHttpInfo(params, headers, callback) {
        if (params === null || params === undefined) {
            throw new Error('Required parameter "params" missing.');
        }
        let data = {};
        data = params;
        if (headers === null || headers === undefined) {
            headers = {};
        }
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let operationVersion = version;
        // CREATE, FETCH, UPDATE operations
        let operationPromise = operationVersion
            .createWithResponseInfo({
            uri: instance._uri,
            method: "post",
            data,
            headers,
        })
            .then((response) => ({
            ...response,
            body: new TranscriptionInstance(operationVersion, response.body),
        }));
        operationPromise = instance._version.setPromiseCallback(operationPromise, callback);
        return operationPromise;
    };
    instance.toJSON = function toJSON() {
        return instance._solution;
    };
    instance[util_1.inspect.custom] = function inspectImpl(_depth, options) {
        return (0, util_1.inspect)(instance.toJSON(), options);
    };
    return instance;
}
